### ISSUE TITLE
Allow to split Java templates into mutliple files

### ISSUE THREAD

**Issue by xylo** [xylo/intellij-postfix-templates#26]

**Comment by thekalinga** [comment]
Also, it would be great if users can import the templates from urls directly.

This way a lot of postfix plugins like mockito & assertj can be subset

**Comment by xylo** [comment]
Yeah.  Separating the templates for different Java libraries in different files was also my intention behind this.  My final aim is to simplify user contributions and the sharing of templates.  However, so far I did not find the perfect solution for this.  But I'm happy to discuss this with someone. :)

I think the solution should allow
- downloading of new template files
- updating (and ideally merging) of template files
- uploading and sharing of template files
- adapting templates that have been downloaded

**Solution 1: URL-based distribution of template files**

Of course downloading and updating could be done by downloading the content of an URL.  But how to merge the new content with eventual local changes?  OK, maybe one can avoid the merging by disallowing changed in those downloaded files.  Instead of editing those templates directly, the user needs to be able to override template definitions in his/her local (offline) files.

This also requires to add a small manager for the URL list to the plugin.

**Solution 2: plugin-based distribution of template files**

Instead of allowing the user to add custom URLs the template files could be contained in the plugin.  However, this creates the restriction that users have to create pull requests for their user contributions.  So this is not optimal, I think.

**Solution 3: VCS-based distribution of template files**

One user wrote me that he manages his template files via version control.  I think this could indeed be a very useful approach because IDEA could probably support us (the user) by managing the version controlled template files directly in IDEA.  Uploads would just be commits in IDEA and downloads/merges could also be done with support of IDEA (diff view).

On the other hand this approach is definitely the most time consuming one, because I have no idea how to tell IDEA to manage one or multiple directories via version control.

**Conclusion**

I really would like to do solution 3 because it's probably the best solution but it could be very hard to implement this.
Thus I guess I will implement the URL-based approach (solution 1).

**Comment by thekalinga** [comment]
Thanks for sharing your thoughts on this

I agree with you that Solution 1 (URL based mechanism) is a reasonable starting point. I have a comprehensive view of how to implement this, including conflict resolution. Here it is

# Basics

We can have a concept called **PostFix Template Source**

A `PostFix Template Source` refers to a physical `postfixTemplates` file that can contains multiple `postfix template`s

For eg., this file
```
.toBoolean : Convert to boolean
	java.lang.String         →  Boolean.parseBoolean($expr$)

.toChar : Convert to character
	java.lang.String         →  $expr$.charAt(0)
```

has two postfix templates with trigger names `toBoolean` & `toChar`

Broadly we will encounter three kinds of **PostFix Template Source**s

1. Editor PostFix Template Source
2. Local File PostFix Template Source
3. Remote PostFix Template Source

User should be allowed to add as many template sources as he chooses

## Editor Template Source

We allow users to Add & Edit templates right inside IDE. These are called `Editor Template Source`. The plugin tracks changes to these files the same way we do now

## Local File Template Source

This refers to a `postfixTemplates` file that comes from the local file system via `file:` protocol. This source is readonly. If the user want to customize any individual template from this source within IDE, he should be allowed to clone & edit the file in IDE

## Remote Template Source

This refers to a `postfixTemplates` file that comes from a http url. This source is readonly too. If the user want to customize any template form this source locally he should be allowed to clone & edit the file in IDE

In both local file & remote template sources, if the source `postfixTemplates` file is changed, we should have an option for the user to refresh all sources. This can be triggered manually by the user

As soon as the user clones from either `Local File Template Source` (or) `Remote Template Source`, the plugin will automatically create a new `Editor Template Source` & adds it the `Editor Template Source` section

If we follow this approach here is how the settings section of this plugin would look like

![custom postfix plugin suggestion](https://user-images.githubusercontent.com/452518/38708320-5e1bef7c-3ed3-11e8-9538-2ece461c4fa6.png)

# Conflict Resolution

Since we have multiple sources, we will bound to have conflicts between each of these sources

We have two kinds of conflicts

1. Inter template type
2. Intra template type

## Inter template type

This refers to conflicts that arise between two types of template sources. For e.g, a single tempalte is defined both in remote template source & in Editor aswell. In this scenario, we can follow a simple rule when it comes to importance

Editor template > Local file template > Remote template

For eg., if we have the following template sources

`Editor Template Source`

```
## number conversions ##

.toBoolean : FROM EDITOR
	java.lang.String         →  Boolean.parseBoolean($expr$)

.toChar : FROM EDITOR
	java.lang.String         →  $expr$.charAt(0)
	NUMBER                   →  ((char) ($expr$))
```

`Local File Template Source`

```
## number conversions ##

.toBoolean : FROM LOCAL FILE
	java.lang.String         →  Boolean.parseBoolean($expr$)

.toByte : FROM LOCAL FILE
	java.lang.String         →  Byte.parseByte($expr$)
```

`Remote Template Source`

```
## number conversions ##

.toBoolean : FROM REMOTE FILE
	java.lang.String         →  Boolean.parseBoolean($expr$)

.toChar : FROM REMOTE FILE
	java.lang.String         →  $expr$.charAt(0)
	NUMBER                   →  ((char) ($expr$))

.toByte : FROM REMOTE FILE
	java.lang.String         →  Byte.parseByte($expr$)
	java.lang.Number         →  $expr$.byteValue()
	NUMBER                   →  ((byte) ($expr$))

.toLong : FROM REMOTE FILE
	java.lang.String         →  Long.parseLong($expr$)
	java.lang.Number         →  $expr$.longValue()
	NUMBER                   →  ((long) ($expr$))
```


The final version that plugin operates would look like the following

```
.toBoolean : FROM EDITOR
	java.lang.String         →  Boolean.parseBoolean($expr$)

.toChar : FROM EDITOR
	java.lang.String         →  $expr$.charAt(0)
	NUMBER                   →  ((char) ($expr$))

.toByte : FROM LOCAL FILE
	java.lang.String         →  Byte.parseByte($expr$)

.toLong : FROM REMOTE FILE
	java.lang.String         →  Long.parseLong($expr$)
	java.lang.Number         →  $expr$.longValue()
	NUMBER                   →  ((long) ($expr$))
```

So the plugin can start reading `Editor Template Source`s first & then add only new templates that are found in `Local File Template Source` followed finally by templates uniquely contributed by `Remote Template Source`

This way, the developer will get the maximum flexibility

## Intra template type

This refers to a scenario where we have multiple templates sources of the same type but have conflicting templates between them. For eg., Lets say there are two `Editor Template Source`s that contribute the same postfix templates, i.e a conflict. This conflict can be resolved based on the order in which sources are declared in the IDE. i.e If there are two `postfix template`s that are conflicting with each other, the one that comes first in the list of sources would take precedence over the other

For eg., let say the two added two `Editor Template Source`s in the IDE with the following templares

`Editor Template Source 1` (Ordered as the first `Editor Template Source` in the settings area of the plugin)

```
## number conversions ##

.toBoolean : FROM EDITOR - FIRST
	java.lang.String         →  Boolean.parseBoolean($expr$)

.toChar : FROM EDITOR - FIRST
	java.lang.String         →  $expr$.charAt(0)
	NUMBER                   →  ((char) ($expr$))
```

`Editor Template Source 2` (Ordered as the second `Editor Template Source` in the settings area of the plugin)

```
## number conversions ##

.toBoolean : FROM EDITOR - SECOND
	java.lang.String         →  Boolean.parseBoolean($expr$)

.toByte : FROM EDITOR - SECOND
	java.lang.String         →  Byte.parseByte($expr$)
```

The final merged version that the plugin sees would look like

```
.toBoolean : FROM EDITOR - FIRST
	java.lang.String         →  Boolean.parseBoolean($expr$)

.toChar : FROM EDITOR - FIRST
	java.lang.String         →  $expr$.charAt(0)
	NUMBER                   →  ((char) ($expr$))

.toByte : FROM EDITOR - SECOND
	java.lang.String         →  Byte.parseByte($expr$)
```

If the user wants to change the priority for any given file, he would need to reorder them using the `Move up` & `Move Down` options in the settings area of the plugin

Please let me know your views

**Comment by xylo** [comment]
@thekalinga 
Thanks for sharing your ideas.  You encouraged me to work on this again.

After thinking about the different lists (editor templates, local templates, remote templates), I think it's not really necessary to manage them in different lists.  I actually like the CheckTree component which IDEA uses to manage the "Postifx Templates" in the settings.  I think I will also use it for the template files.  The different types (local, remote, ...) could just be highlighted differently.  Moreover this common management of remote and local template files has the advantage that you are more flexible in prioritizing the template files and I think this is needed for a good template composition.

### Template Composition

Let's take the following example:

Template file 1:
```
.content : get content
	java.io.File        →  new String(java.nio.file.Files.readAllBytes($expr$.toPath()), $encoding::"\"UTF-8\""$)
```

Template file 2:
```
.content : get content
	java.nio.file.Path  →  new String(java.nio.file.Files.readAllBytes($expr$), $encoding::"\"UTF-8\""$)
```

Template file 3:
```
.content : get content
	java.lang.Object    →  contentOf($expr$)
```

Each of these files defines a rule for the `.content` template.  While the rules of the first two files match against distinct types, the more general rule in the third file can be also applied when the first two rules can be applied.  So in this case we would like the rule of the third file to be processed last.  Or in other words we want the template composition algorithm to create a template that operates like:
```
.content : get content
	java.io.File        →  new String(java.nio.file.Files.readAllBytes($expr$.toPath()), $encoding::"\"UTF-8\""$)
	java.nio.file.Path  →  new String(java.nio.file.Files.readAllBytes($expr$), $encoding::"\"UTF-8\""$)
	java.lang.Object    →  contentOf($expr$)
```
Therefore, the third file has to be on the third position in this case.

### Disabling rules

Apart from the possibility to disable a whole template file by unchecking it in the tree there should be also a way to disable a single rule (e.g. `.content` on `java.lang.Object`).  I'm sill not sure about the best way to do this, but I see at the moment two options for this:
- We can either allow blocking rules that reject the template application on certain types, e.g.:
```
	java.lang.Object    →  DISABLED
```
Those rules have to be placed prior to the rule to disable.
- Another option would be to allow for subsequent deletion of rules, e.g. if template file 3 defines a rule on `java.lang.Object` I can remove this rule in template file 4 via
```
	java.lang.Object    →  REMOVE
```
or something like that.

Both approaches have IMO pros & cons.  The advantage of the DISABLED rule is that we always follow the _first come first served_ concept.  However, it does not allow to remove or skip a rule so that the subsequent rules can still be applied.  And this is exactly the advantage of the REMOVE rule.  So I'm not sure about this.  Maybe I implement both of them.

**Comment by thekalinga** [comment]
Yes.. I missed template merging section completely

Having all of types of templates in a single place would make sense only if there are scenarios where they needs to be mixed and matched

Plz correct me if I'm wrong. When there are local templates, they are always more important the the remote sources. Editor level changes are always more important than the local changes

I assume you would be taking care of this priority hierarchy out of the box

Thanks in anycase for sharing the details

**Comment by xylo** [comment]
I think it will be quite common that the same template name is used in multiple files with different rules.  Take for instance the `.content` template which represents the concept of _getting the content of something_ and which can be applied to files, URLs, streams, and so on.  So there are a lot of Java classes that can be enriched by this template, especially when you think about classes of external libraries.  So there is definitely the need for combining template rules of different files.

Moreover I think it's not necessary and not advisable to enforce a strict priority policy on template types.  For instance, let's say we have an web template file for IDEA plugin developer templates.  This file contains a `.content` template for the class `VirtualFile`.  Moreover, I created a `.content` template for the class `java.lang.Object` in my local template file.  If my local template file has a higher priority than the web template file, the template rule for `VirtualFile` is never applied.  So in this case my file needs to be placed below the web template file.

However, there are also cases were I want certain template rules of mine to come first.  So in general it should be possible to create different local template files and give them different priorities (order them above or below web templates).  And this will be possible with the tree implementation.

**Comment by thekalinga** [comment]
@xylo Agree with you that the merging needs to happen. But I'm talking about the sub-template that is in conflict

i.e

`Form local`
```
.content : LOCAL DESCRIPTION
	java.io.File        →  LOCAL IMPLEMENTATION
```

`Form Remote`
```
.content : REMOTE DESCRIPTION
	java.io.File        →  REMOETE IMPLEMENTATION
	java.nio.file.Path  →  new String(java.nio.file.Files.readAllBytes($expr$), $encoding::"\"UTF-8\""$)
	java.lang.Object    →  contentOf($expr$)
```

My assumption is that the final version that the plugin would operate on (after merging) would be

```
.content : LOCAL DESCRIPTION
	java.io.File        →  LOCAL IMPLEMENTATION
.content : REMOTE DESCRIPTION
	java.nio.file.Path  →  new String(java.nio.file.Files.readAllBytes($expr$), $encoding::"\"UTF-8\""$)
	java.lang.Object    →  contentOf($expr$)
```

In the above example, the `sub template` `.content` for `File` is in conflict. So, the local suggestion should always take precedence over the remote one

Please note I have added different descriptions for each templates aswell

**Comment by xylo** [comment]
I would say the general rule is: _first come first served_

So regarding your example and in the case that you place your local template file above the remote template file, the resulting template is
```
.content : LOCAL DESCRIPTION
	java.io.File        →  LOCAL IMPLEMENTATION
        ADDITIONAL_LOCAL    →  ADDITIONAL LOCAL IMPLEMENTATION
	java.nio.file.Path  →  new String(java.nio.file.Files.readAllBytes($expr$), $encoding::"\"UTF-8\""$)
	java.lang.Object    →  contentOf($expr$)
```

**Comment by xylo** [comment]
Unfortunately, there can be only one template description.  Thus the first description wins, I would say.

**Comment by thekalinga** [comment]
I dont know if First come First Served is a good default, as we always know for a fact user writes his own templates because he wants to override the remote ones

**Comment by xylo** [comment]
Don't worry, with the flexible tree-based solution the user will be able to augment remote sub-templates and do everything he wants.

**Comment by xylo** [comment]
I finally released v2.0.
